### PR TITLE
bun-baseline: init at 0.7.0 

### PR DESCRIPTION
**Pull request by neowsl**
###### Description of changes

`bun-baseline` is a build of the bun runtime designed for CPUs without AVX2 instructions
Homepage: https://github.com/oven-sh/bun
Most of the code is from https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/development/web/bun/default.nix

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by necrophcodr** [comment]
@mathletedev This does not conform to the CONTRIBUTING.md guidelines. Even as of the time of making the PR, it seems the commits are incorrectly formatted.

File: pkgs/development/web/bun-baseline/default.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  version = "0.7.0";
+  pname = "bun-baseline";
+
+  src = fetchurl {
+    url = "https://github.com/Jarred-Sumner/bun-releases-for-updater/releases/download/bun-v${version}/bun-${dist.shortName}-${dist.arch}-baseline.zip";
+    sha256 = dist.sha256;
+  };
+
+  strictDeps = true;
+  nativeBuildInputs = [ unzip ] ++ lib.optionals stdenvNoCC.isLinux [ autoPatchelfHook ];
```
Comment (SuperSandro2000) [inline pkgs/development/web/bun-baseline/default.nix:28]: we want to build from source

**Comment by mathletedev** [comment]
@SuperSandro2000 Sorry about the trouble; this is my first time contributing. Does the new commit work?

**Comment by SuperSandro2000** [comment]
yeah, that looks good

**Comment by necrophcodr** [comment]
> @SuperSandro2000 Sorry about the trouble; this is my first time contributing. Does the new commit work?

Are these builds not still from binary packaging? Is it not possible to perform the baseline build?

**Comment by mathletedev** [comment]
My bad, I assumed the nix script from the non-baseline Bun package built from source. I use the packaged releases of Bun on my own machine, which has worked great for development, but I haven't really dabbed in building it. I don't have the experience, so if it is necessary to build this from source, I can close this PR. I apologise for the confusion. Have a wonderful day!

**Comment by drupol** [comment]
@GrahamcOfBorg build bun-baseline

**Comment by SuperSandro2000** [comment]
> does it make sense to require this to be built from source?

Yes, if it is possible we want to build from source.

> It seems to me perhaps that in the future, the Bun derivation should instead be setup to produce a baseline and non-baseline derivation, since the build process for the two might be very comparable.

I am not sure what do you mean.

**Review by drupol — CHANGES_REQUESTED** [review]
You need to add a `bun-baseline` entry in `all-packages.nix`.

**Comment by necrophcodr** [comment]
> It seems to me perhaps that in the future, the Bun derivation should instead be setup to produce a baseline and non-baseline derivation, since the build process for the two might be very comparable.

I'm  saying that bun (the existing derivation) and bun-baseline (this derivation) would both build the exact same codebase. Currently neither of them build anything, they just download and install a prebuilt binary executable. The reason for bun-baseline to exist, is so that bun can run on earlier x86-64 CPUs, since the most recent compiled version would be using a set of features that older CPUs would not support.

Having two distinct derivations seems a bit overkill to me, when in reality it's (more or less) a matter of building bun with zig normally, or with the CPU model specified as `baseline`.  To be clear, I have not looked into the bun-baseline project to see if there are any additional patches applied to it, but given that the `bun` upstream codebase already has checks for this baseline compilation, I would expect not.

**Comment by mathletedev** [comment]
> Having two distinct derivations seems a bit overkill to me, when in reality it's (more or less) a matter of building bun with zig normally, or with the CPU model specified as baseline. To be clear, I have not looked into the bun-baseline project to see if there are any additional patches applied to it, but given that the bun upstream codebase already has checks for this baseline compilation, I would expect not.

That sounds like a great idea! I initially wanted to modify the existing bun derivation, but I didn't know how to check the CPU model during compilation. If someone is more experienced with Nix, I think that approach would be better.

**Comment by necrophcodr** [comment]
> > Having two distinct derivations seems a bit overkill to me, when in reality it's (more or less) a matter of building bun with zig normally, or with the CPU model specified as baseline. To be clear, I have not looked into the bun-baseline project to see if there are any additional patches applied to it, but given that the bun upstream codebase already has checks for this baseline compilation, I would expect not.
> 
> That sounds like a great idea! I initially wanted to modify the existing bun derivation, but I didn't know how to check the CPU model during compilation. If someone is more experienced with Nix, I think that approach would be better.

You may need to do some experimentation with this, but as far as I can tell, it's a matter of using `zig build` vs `zig build -Dcpu=baseline`. I could well be wrong though, as I have no real way to test if that would produce a binary that runs on older hardware.

From upstream build.zig:

```
...
        const is_baseline = arch.isX86() and (target.cpu_model == .baseline or
            !std.Target.x86.featureSetHas(target.getCpuFeatures(), .avx2));
...
```

**Comment by renhiyama** [comment]
Any update on this? I really need bun natively on my old pc 😅

**Comment by necrophcodr** [comment]
> Any update on this? I really need bun natively on my old pc 😅

You can still derive from the source commit that this PR is based upon and get the binary variant. But it won't exist for any longer than this PR, if that.

@mathletedev have you had the option of looking into this since?

**Comment by DamienCassou** [comment]
Should we close this PR?

**Pull request closed without merging**